### PR TITLE
changes on PR 69 + bug fix

### PR DESCRIPTION
**Pull request by b-biswas**
made changes based on comments on #69.
Also fixed a couple of bugs in the deblend functions

### CODE REVIEW

File: debvader/deblender.py [hunk 1]
```diff
@@ -555,12 +559,14 @@ def deblending_step(
 
         paramters:
```
Comment (aboucaud) [inline debvader/deblender.py:560]:
```suggestion
        parameters:
```

File: debvader/deblender.py [hunk 2]
```diff
@@ -34,6 +34,7 @@ def __init__(
         nb_of_bands=6,
         epistemic_uncertainty_estimation=False,
         normalization_func=None,
+        denormalization_func=None,
```
Comment (aboucaud) [inline debvader/deblender.py:37]: Alright, this is adding too many arguments for one simple task to be handled.
Plus I guess you have to have to ensure identical normalisation functions are used, either linear on non-linear. 
Let's try to find an alternative..

Two options come to mind : 
- use a class e.g. `LinearCosmos` and `NonLinearCosmos` with methods `normalize()` and `denormalize()`
- use more specific keywords like `normalization_type="linear"` and `normalization_survey="cosmos"`, and from these, go fetch the right function accordingly.

Comment (b-biswas) [reply debvader/deblender.py]: I agree with the point there should be only one function. But I want to give the user the flexibility to choose easily define the normalization function. 
I just push changes to the PR, does it look ok?

File: debvader/batch_generator.py [hunk 3]
```diff
@@ -24,21 +26,27 @@ def __init__(
...
         self.list_of_samples = list_of_samples
         self.x_col_name = x_col_name
         self.y_col_name = y_col_name
         self.batch_size = batch_size
         self.num_iterations_per_epoch = num_iterations_per_epoch
-        self.normalization_func = normalization_func
+        if normalization is not None:
+            if isinstance(normalization, Normalization):
+                raise ValueError(
+                    "The parameter `normalization` shoudl be an instance of Debvader.normalize.Normalization"
```
Comment (aboucaud) [inline debvader/batch_generator.py:40]:
```suggestion
                    "The parameter `normalization` should be an instance of Debvader.normalize.Normalization"
```

File: debvader/batch_generator.py [hunk 4]
```diff
@@ -24,21 +26,27 @@ def __init__(
...
         channel_last: boolean to indicate if the the clast channel corresponds to differnet bands of the input data.
         """
         self.list_of_samples = list_of_samples
         self.x_col_name = x_col_name
         self.y_col_name = y_col_name
         self.batch_size = batch_size
         self.num_iterations_per_epoch = num_iterations_per_epoch
-        self.normalization_func = normalization_func
+        if normalization is not None:
+            if isinstance(normalization, Normalization):
```
Comment (aboucaud) [inline debvader/batch_generator.py:38]:
```suggestion
            if not isinstance(normalization, Normalization):
```

Comment (aboucaud) [reply debvader/batch_generator.py:38]: This could actually be written in a single line as `if not ... and not ...:`

File: debvader/batch_generator.py [hunk 5]
```diff
@@ -24,21 +26,27 @@ def __init__(
...
+        normalization: object of Debvader.normalize.Normalize, used to perform norm and denorm operations
         channel_last: boolean to indicate if the the clast channel corresponds to differnet bands of the input data.
         """
         self.list_of_samples = list_of_samples
         self.x_col_name = x_col_name
         self.y_col_name = y_col_name
         self.batch_size = batch_size
         self.num_iterations_per_epoch = num_iterations_per_epoch
-        self.normalization_func = normalization_func
+        if normalization is not None:
```
Comment (aboucaud) [inline debvader/batch_generator.py:37]: If you make that check, then I guess the default value for `normalization` should be `None`

File: debvader/normalize.py [hunk 6]
```diff
@@ -1,32 +1,61 @@
...
 
-def linear_normalization_cosmos(x, direction="normalize"):
+class Normalization(ABC):
+    @abstractmethod
+    def forward(self, images):
+        pass
+
+    @abstractmethod
+    def inverse(self, images):
+        pass
```
Comment (aboucaud) [inline debvader/normalize.py:13]: Looks good. 

Few comments on the naming.
- the object itself is not a normalization, but rather a Normalizer. I would call it this way so it is clear
- as I mentioned earlier, both operations should have names that indicate the opposite. It is not clear for me that `inverse` is the opposite of `forward`. It suggest to use either `forward/backward` or the same api as scikit-learn: `transform/inverse_transform`.

File: debvader/deblender.py [hunk 7]
```diff
@@ -53,7 +54,12 @@ def __init__(
         self.cutout_size = cutout_size
         self.nb_of_bands = nb_of_bands
         self.epistemic_uncertainty_estimation = epistemic_uncertainty_estimation
-        self.normalization_func = normalization_func
+        if normalization is not None:
+            if isinstance(normalization, Normalization):
```
Comment (aboucaud) [inline debvader/deblender.py:58]: here also I guess there should be a `not`

File: debvader/batch_generator.py (hunk 5 above)
Comment (b-biswas) [reply debvader/batch_generator.py:37]: Yes, by default there was no normalization while training the dc2 simulations

**Review by aboucaud — COMMENTED** [review]
changes in the naming should also be ported to `debvader/batch_generator.py` (still using `Normalization`)

File: debvader/normalize.py (hunk 6 above)
Comment (b-biswas) [reply debvader/normalize.py]: done!

File: debvader/batch_generator.py (hunk 5 above)
Comment (aboucaud) [reply debvader/batch_generator.py:37]: I mean in the `__init__`, it should be a keyword with default value `None`

File: debvader/batch_generator.py [hunk 8]
```diff
@@ -3,6 +3,8 @@
 import numpy as np
 from tensorflow.keras.utils import Sequence
 
+from debvader.normalize import Normalization
```
Comment (aboucaud) [inline debvader/batch_generator.py:6]:
```suggestion
from debvader.normalize import Normalizer
```

File: debvader/batch_generator.py [hunk 9]
```diff
@@ -24,21 +26,27 @@ def __init__(
         y_col_name: column name of data to be fed as target to the network
         batch_size: sample sixe for each batch
         num_iterations_per_epoch: number of samples (of size = batch_size) to be drawn from the sample
-        normalization_func: function to be used for normalization
+        normalization: object of Debvader.normalize.Normalize, used to perform norm and denorm operations
```
Comment (aboucaud) [inline debvader/batch_generator.py:29]:
```suggestion
        normalization: instance of Debvader.normalize.Normalizer, used to perform forward and backward normalization operations
```

File: debvader/batch_generator.py [hunk 10]
```diff
@@ -12,7 +14,7 @@ def __init__(
         y_col_name,
         batch_size,
         num_iterations_per_epoch,
-        normalization_func,
+        normalization,
```
Comment (aboucaud) [inline debvader/batch_generator.py:17]:
```suggestion
        normalization=None,
```

File: debvader/batch_generator.py [hunk 11]
```diff
@@ -26,22 +26,20 @@ def __init__(
...
         self.y_col_name = y_col_name
         self.batch_size = batch_size
         self.num_iterations_per_epoch = num_iterations_per_epoch
-        if (normalization is not None) and (
-            not isinstance(normalization, Normalization)
-        ):
+        if (normalizer is not None) and (not isinstance(normalizer, Normalizer)):
             raise ValueError(
-                "The parameter `normalization` should be an instance of Debvader.normalize.Normalization"
+                "The parameter `normalizer` should be an instance of Debvader.normalize.Normalizer"
```
Comment (aboucaud) [inline debvader/batch_generator.py:39]:
```suggestion
                "The parameter `normalizer` should be an instance of debvader.normalize.Normalizer"
```

**Review by aboucaud — APPROVED** [review]

**Comment by aboucaud** [comment]
Ready for merging as soon as the CI passes
